### PR TITLE
SAFE v2

### PR DESCRIPTION
**Pull request by kerzhner**
Based on the [design doc](https://www.notion.so/statechannels/HTLC-free-SAFE-e999aafe604c4ce3bc69aec1000971b0?d=f803b78c4de442a49fa6eb069f387733#020d60ae8f054a4589bbdefd88503880)

This PR includes:
- Happy path solidity code.
- A test to verify that the happy path works as expected.

### CODE REVIEW

File: contracts/l1.sol [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            recoverSigner(message, signature) == lpAddress,
+            "Signed by liquidity provider"
+        );
+        require(
+            swaps[0].timestamp + l1ClaimWindow > block.timestamp,
+            "Within claim window"
+        );
+
+        for (uint256 i = 0; i < swaps.length; i++) {
+            swaps[i].l1Recipient.call{value: swaps[i].value}("");
```
Comment (lalexgap) [inline contracts/l1.sol:29]: My knowledge is probably out of date but I think this won't revert if it fails?
I think we need

```
     (bool sent, bytes memory data) = _to.call{value: msg.value}("");
        require(sent, "Failed to send Ether");
```
based on https://solidity-by-example.org/sending-ether/?

File: contracts/l2.sol [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+// TODO: update these values after prototype phase
+uint256 constant authorizationWindow = 60;
+uint256 constant l2ClaimWindow = 180;
+
+contract L2 is SignatureChecker {
+    RegisteredSwap[] public registeredSwaps;
+    // `batches` is used to record the fact that tickets with nonce
+    // between startingNonce and startingNonce + numSwaps-1 are
+    // authorized, claimed or returned.
+    mapping(uint256 => Batch) batches;
```
Comment (lalexgap) [inline contracts/l2.sol:44]: Is this indexed by nonce?

File: contracts/l2.sol [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        );
+        uint256 earliestTimestamp = registeredSwaps[first].timestamp;
+
+        require(
+            recoverSigner(message, signature) == lpAddress,
+            "Signed by liquidity provider"
+        );
+        require(
+            earliestTimestamp + authorizationWindow > block.timestamp,
+            "Within autorization window"
```
Comment (lalexgap) [inline contracts/l2.sol:93]: Nit but I think the convention is that `require` message is an error message? Maybe something like?
```
require( earliestTimestamp + authorizationWindow > block.timestamp, "Swap not within authorization window");
```

File: test/safe.test.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  const signature = signData(hashSwaps(swapsWithIndex), lpPK);
+  await lpL2.authorizeWithdrawal(
+    0,
+    0,
+    signData(hashSwaps(swapsWithIndex), lpPK),
+  );
+  await lpL1.claimBatch([swap], signature);
+
+  await ethers.provider.send("evm_increaseTime", [121]);
+  await lpL2.claimL2Funds(0);
```
Comment (lalexgap) [inline test/safe.test.ts:51]: Could we expand this to check if Alice/Bob's balances change as expected?

File: hardhat.config.ts [hunk 5]
```diff
@@ -25,12 +25,12 @@ const config: HardhatUserConfig = {
       accounts: [
         {
           privateKey:
-            "0x91f47a1911c0fd985b34c25962f661f0de606f7ad38ba156902dff48b4d05f97",
+            "0xe1743f0184b85ac1412311be9d6e5d333df23e22efdf615d0135ca2b9ed67938",
```
Comment (lalexgap) [inline hardhat.config.ts:28]: Is there a reason to change the funded private keys?

File: contracts/l1.sol [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        }
+
+        latestNonceWithdrawn = latestNonceWithdrawn + swaps.length; // maybe off by one;
```
Comment (lalexgap) [inline contracts/l1.sol:32]: Did we figure out of this is `off by one` ?

File: contracts/l2.sol [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            value: deposit.depositAmount,
+            timestamp: block.timestamp
+        });
+
+        // swap's nonce is now its index in `registeredSwaps`
+        registeredSwaps.push(swap);
+    }
+
+    function authorizeWithdrawal(
+        uint256 first,
```
Comment (lalexgap) [inline contracts/l2.sol:70]: Do we have validate the `first` or `last` values in some way?

File: contracts/l1.sol (hunk 1 above)
Comment (lalexgap) [reply contracts/l1.sol:29]: Locally adding the require statement I see this error:
```
     Error: cannot estimate gas; transaction may fail or may require manual gas limit (error={"stackTrace":[{"type":18,"address":{"type":"Buffer","data":[197,251,165,48,223,166,33,238,173,24,119,197,237,7,116,184,230,60,245,153]},"message":{"value":{"type":"Buffer","data":[8,195,121,160,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,32,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,20,70,97,105,108,101,100,32,116,111,32,115,101,110,100,32,69,116,104,101,114,0,0,0,0,0,0,0,0,0,0,0,0]},"_selector":"08c379a0"}}]}, tx={"data":"0x22dc0ede0000000000000000000000000000000000000000000000000000000000000080c5060b11f205efaed88cb836aab9490d4a05ddde59465a49dedc92f51a0b73d135dc6f8100fa99a669afdb947c8f8eff4d864ab40bdf37e9d58bd954bca272ce000000000000000000000000000000000000000000000000000000000000001b00000000000000000000000000000000000000000000000000000000000000010000000000000000000000002a47cd5718d67dc81eafb24c99d4db159b0e7bca00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000061b150df","to":{},"from":"0x2a47Cd5718D67Dc81eAfB24C99d4db159B0e7bCa","type":2,"maxFeePerGas":{"type":"BigNumber","hex":"0xdbbc4554"},"maxPriorityFeePerGas":{"type":"BigNumber","hex":"0x9502f900"},"nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.5.0)
  ```

File: contracts/l2.sol (hunk 3 above)
Comment (lalexgap) [reply contracts/l2.sol:93]: Oops I think I just misread this.

File: contracts/l2.sol [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            total: total,
+            earliestTimestamp: earliestTimestamp,
+            latestTimestamp: registeredSwaps[last].timestamp,
+            status: BatchStatus.Pending
+        });
+    }
+
+    function claimL2Funds(uint256 first) public {
+        Batch memory batch = batches[first];
+        require(batch.status == BatchStatus.Pending, "Batch is pending");
```
Comment (lalexgap) [inline contracts/l2.sol:107]: Nit but I think convention is that the `require` message is the error case?  Maybe something like:
```
require(batch.status == BatchStatus.Pending, "The batch must be in a pending status");
```

Comment (kerzhner) [reply contracts/l2.sol]: Good call! Fixed in a402372

File: contracts/l2.sol (hunk 7 above)
Comment (kerzhner) [reply contracts/l2.sol:73]: I think we do! In the latest design doc, there is a suggestion that batches should be non-overlapping. I'll insert a TODO and leave that work for future PRs.

File: contracts/l1.sol (hunk 6 above)
Comment (kerzhner) [reply contracts/l1.sol]: I removed the comment in 068a99d. But more tests are needed to verify that the code works as expected.

File: hardhat.config.ts (hunk 5 above)
Comment (kerzhner) [reply hardhat.config.ts:28]: No real reason. I just started using new keys before I saw the definitions for Alice + Bob private keys.

File: test/safe.test.ts (hunk 4 above)
Comment (kerzhner) [reply test/safe.test.ts:51]: We should! We need to add a lot more testing around the code in general.

My goal with this PR is to translate the design doc happy-path pseudo-code into solidity + add a sanity check unit test. With this PR, future work can be unblocked and parallelized.

File: contracts/l2.sol (hunk 2 above)
Comment (kerzhner) [reply contracts/l2.sol:43]: Yep! Comment clarified in f16ac14

File: contracts/l1.sol (hunk 1 above)
Comment (kerzhner) [reply contracts/l1.sol]: Good catch. Fixed in c22ae65. But this code will need to be modified as soon as ERC20 support is added.

**Review by lalexgap — APPROVED** [review]

**Pull request merged**